File: Week 02/id_158/144.二叉树的前序遍历.cs
/*
 * @lc app=leetcode.cn id=144 lang=csharp
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> PreorderTraversal(TreeNode root) {
        IList<int> res = new List<int>();
        Preorder(root,res);
        return res;
    }

    private void Preorder(TreeNode treeNode,IList<int> res){
        if(treeNode==null) return;
        res.Add(treeNode.val);
        Preorder(treeNode.left,res);
        Preorder(treeNode.right,res);
    }

    public IList<int> PreorderTraversalWithStack(TreeNode root){
        IList<int> res = new List<int>();
        if(root==null){
            return res;
        }
        TreeNode treeNode = root;
        Stack<TreeNode> stack = new Stack<TreeNode>();
        while (treeNode!=null || stack.Count>0)
        {
            while (treeNode!=null)
            {
                res.Add(treeNode.val);
                stack.Push(treeNode);
                treeNode = treeNode.left;
            }
            if(stack.Count>0){
                treeNode=stack.Pop();
                treeNode = treeNode.right;
            }
        }
        return res;
    }
}
// @lc code=end

File: Week 02/id_493/144.二叉树的前序遍历.js
/*
 * @lc app=leetcode.cn id=144 lang=javascript
 *
 * [144] 二叉树的前序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    if (root == null) {
        return [];
    }

    let result = [];
    let stArr = [root];
    while (stArr.length > 0) {
        let ff = stArr.pop();
        result.push(ff.val);
        if (ff.right !== null) {
            stArr.push(ff.right);
        }
        if (ff.left !== null) {
            stArr.push(ff.left);
        }
    }
    return result;

    // let result = [];
    // result.push(root.val);

    // let left = postorderTraversal(root.left);
    // let right = postorderTraversal(root.right);
    // result.push(...left);
    // result.push(...right);
    
    // return result;

    // let result = [];
    // let stArr = [];
    // while (root != null || stArr.length > 0) {
    //     while (root !== null) {
    //         stArr.push(root);
    //         result.push(root.val);
    //         root = root.left;
    //     }
    //     root = stArr.pop();
    //     root = root.right;
    // }
    // return result;
};
// @lc code=end

